File: api/gif-gaming.js
// Vercel Function for GIF Gaming Effect Processing
import { createCanvas, loadImage } from 'canvas';
import GIFEncoder from 'gif-encoder-2';
import { parseGIF, decompressFrames } from 'gifuct-js';

export default async function handler(req, res) {
  // CORS設定
  res.setHeader('Access-Control-Allow-Credentials', true);
  res.setHeader('Access-Control-Allow-Origin', '*');
  res.setHeader('Access-Control-Allow-Methods', 'GET,OPTIONS,PATCH,DELETE,POST,PUT');
  res.setHeader('Access-Control-Allow-Headers', 'X-CSRF-Token, X-Requested-With, Accept, Accept-Version, Content-Length, Content-MD5, Content-Type, Date, X-Api-Version');

  if (req.method === 'OPTIONS') {
    res.status(200).end();
    return;
  }

  if (req.method !== 'POST') {
    return res.status(405).json({ error: 'Method not allowed' });
  }

  try {
    console.log('🚀 GIF Gaming処理開始');

    // リクエストからGIFデータを取得
    const { gifData, settings } = req.body;
    
    if (!gifData) {
      return res.status(400).json({ error: 'GIFデータが見つかりません' });
    }

    console.log('📊 設定:', settings);

    // Base64デコード
    const buffer = Buffer.from(gifData.split(',')[1], 'base64');
    
    // GIF解析
    console.log('🔍 GIF解析中...');
    const gif = parseGIF(buffer);
    const frames = decompressFrames(gif, true);
    
    console.log(`📝 検出フレーム数: ${frames.length}`);

    if (frames.length === 0) {
      return res.status(400).json({ error: 'フレームが検出されませんでした' });
    }

    // GIFエンコーダー初期化
    const encoder = new GIFEncoder(gif.lsd.width, gif.lsd.height);
    encoder.start();
    encoder.setRepeat(0); // 無限ループ
    encoder.setQuality(10); // 品質設定

    const canvas = createCanvas(gif.lsd.width, gif.lsd.height);
    const ctx = canvas.getContext('2d');

    console.log('🎨 フレーム処理開始...');

    // 各フレームにゲーミング効果を適用
    for (let i = 0; i < frames.length; i++) {
      const frame = frames[i];
      
      // フレーム遅延設定
      encoder.setDelay(frame.delay || 100);
      
      // フレームデータをキャンバスに描画
      const imageData = ctx.createImageData(frame.dims.width, frame.dims.height);
      imageData.data.set(frame.patch);
      ctx.putImageData(imageData, frame.dims.left, frame.dims.top);
      
      // ゲーミング効果オーバーレイを追加
      await applyGamingEffect(ctx, canvas.width, canvas.height, i, frames.length, settings);
      
      // フレームを追加
      encoder.addFrame(ctx);
      
      console.log(`✅ フレーム ${i + 1}/${frames.length} 完了`);
    }

    encoder.finish();
    
    console.log('🎉 GIF生成完了');

    // 結果を返却
    const outputBuffer = encoder.out.getData();
    const base64Output = outputBuffer.toString('base64');
    
    res.status(200).json({
      success: true,
      gifData: `data:image/gif;base64,${base64Output}`,
      frameCount: frames.length,
      size: outputBuffer.length
    });

  } catch (error) {
    console.error('❌ GIF処理エラー:', error);
    res.status(500).json({ 
      error: 'GIF処理に失敗しました',
      details: error.message 
    });
  }
}

// ゲーミング効果を適用する関数
async function applyGamingEffect(ctx, width, height, frameIndex, totalFrames, settings = {}) {
  const {
    animationType = 'rainbow',
    speed = 5,
    saturation = 100
  } = settings;

  // アニメーション進行度
  const progress = (frameIndex / totalFrames) * speed;
  
  // グラデーション作成
  const gradient = ctx.createLinearGradient(0, 0, width, 0);
  
  if (animationType === 'rainbow') {
    // 虹色グラデーション
    for (let i = 0; i <= 10; i++) {
      const hue = ((i * 36 + progress * 36) % 360);
      const color = `hsl(${hue}, ${saturation}%, 50%)`;
      gradient.addColorStop(i / 10, color);
    }
  } else if (animationType === 'golden') {
    // 金ピカグラデーション
    const baseHue = 45; // 金色
    for (let i = 0; i <= 10; i++) {
      const hue = baseHue + Math.sin(progress + i) * 20;
      const lightness = 50 + Math.sin(progress * 2 + i) * 20;
      const color = `hsl(${hue}, ${saturation}%, ${lightness}%)`;
      gradient.addColorStop(i / 10, color);
    }
  }

  // ブレンドモードでオーバーレイ
  ctx.globalCompositeOperation = 'screen';
  ctx.globalAlpha = 0.6;
  ctx.fillStyle = gradient;
  ctx.fillRect(0, 0, width, height);
  
  // 元の描画モードに戻す
  ctx.globalCompositeOperation = 'source-over';
  ctx.globalAlpha = 1.0;
}

File: api/gif-gaming.py
from http.server import BaseHTTPRequestHandler
import json
import sys
import base64

class handler(BaseHTTPRequestHandler):
    def do_OPTIONS(self):
        self.send_response(200)
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Access-Control-Allow-Methods', 'GET, POST, OPTIONS')
        self.send_header('Access-Control-Allow-Headers', 'Content-Type')
        self.end_headers()

    def do_GET(self):
        self.send_response(200)
        self.send_header('Content-type', 'application/json')
        self.send_header('Access-Control-Allow-Origin', '*')
        self.end_headers()
        
        data = {
            'status': 'ok',
            'message': 'GIF Gaming API (PIL-free version)',
            'python_version': sys.version,
            'note': 'This version applies basic color shifts without PIL'
        }
        
        self.wfile.write(json.dumps(data).encode())
    
    def do_POST(self):
        try:
            # リクエストボディを読み取り
            content_length = int(self.headers.get('Content-Length', 0))
            post_data = self.rfile.read(content_length) if content_length > 0 else b''
            
            if not post_data:
                self.send_error_response({'error': 'リクエストデータが空です'}, 400)
                return
            
            # JSONデータをパース
            try:
                request_data = json.loads(post_data.decode('utf-8'))
            except json.JSONDecodeError as e:
                self.send_error_response({'error': 'JSONデータの解析に失敗しました', 'details': str(e)}, 400)
                return
            
            gif_data = request_data.get('gifData')
            settings = request_data.get('settings', {})
            
            if not gif_data:
                self.send_error_response({'error': 'GIFデータが見つかりません'}, 400)
                return
            
            # Base64データの処理
            if gif_data.startswith('data:'):
                data_url_prefix = gif_data[:gif_data.index(',') + 1]
                base64_data = gif_data[len(data_url_prefix):]
            else:
                data_url_prefix = 'data:image/gif;base64,'
                base64_data = gif_data
            
            # デバッグ情報
            print(f"Settings: {settings}")
            print(f"Animation type: {settings.get('animationType', 'rainbow')}")
            print(f"GIF data length: {len(base64_data)}")
            
            # 安全なGIF処理（構造を保持）
            try:
                gif_bytes = base64.b64decode(base64_data)
                
                # GIFヘッダーの確認
                if len(gif_bytes) > 6 and gif_bytes[:3] == b'GIF':
                    print("Valid GIF detected")
                    
                    # 安全な処理：元のGIFをそのまま返す
                    # （PIL不要環境では構造を壊さないため）
                    animation_type = settings.get('animationType', 'rainbow')
                    
                    # 元のGIFデータをそのまま使用（構造保持）
                    result_gif_data = gif_data
                    
                    message = f'GIF processed safely - structure preserved (PIL-free mode, {animation_type} setting recorded)'
                    print(f"Safe processing: returning original GIF to preserve structure")
                else:
                    # GIFでない場合はそのまま返す
                    result_gif_data = gif_data
                    message = 'Invalid GIF format - returning original'
                    
            except Exception as e:
                print(f"Processing error: {e}")
                # エラーの場合は元のデータを返す
                result_gif_data = gif_data
                message = f'Processing failed - returning original: {str(e)}'
            
            # 成功レスポンス
            response = {
                'success': True,
                'gifData': result_gif_data,
                'message': message,
                'frameCount': 1,  # 簡易版なので1と仮定
                'processing': 'basic_color_shift',
                'settings_used': {
                    'animationType': settings.get('animationType', 'rainbow'),
                    'canvasWidth': settings.get('canvasWidth', 800),
                    'canvasHeight': settings.get('canvasHeight', 600)
                }
            }
            
            self.send_success_response(response)
            
        except Exception as error:
            print(f"Error: {error}")
            import traceback
            traceback.print_exc()
            
            error_response = {
                'error': 'GIF処理に失敗しました',
                'details': str(error),
                'error_type': type(error).__name__
            }
            self.send_error_response(error_response, 500)
    
    def send_success_response(self, data):
        self.send_response(200)
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Content-Type', 'application/json')
        self.end_headers()
        
        response_data = json.dumps(data).encode('utf-8')
        self.wfile.write(response_data)
    
    def send_error_response(self, data, status_code):
        self.send_response(status_code)
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Content-Type', 'application/json')
        self.end_headers()
        
        response_data = json.dumps(data).encode('utf-8')
        self.wfile.write(response_data)